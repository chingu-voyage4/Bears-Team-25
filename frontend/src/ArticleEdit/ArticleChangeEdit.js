import React from 'react';
import PropTypes from 'prop-types';

import PrimaryButton from '../common/PrimaryButton';
import MessageBar from '../common/MessageBar';

import actions from './actions';

import { Heading1, Wrapper } from './styled';

// import { SMALL_WINDOW } from '../config';

/* eslint-disable camelcase */

export default class ArticleChangeEdit extends React.Component {
  static propTypes = {
    id: PropTypes.string,
  };
  static defaultProps = {
    id: '',
  };

  state = {
    changeRequest: [],
    horizontal: 'right',
    vertical: 'top',
    message: { show: false, error: false, text: '' },
  };

  componentDidMount() {
    this.fetchData(this.props.id).then(results => {
      this.setState({ changeRequest: results });
    });
  }

  // eslint-disable-next-line class-methods-use-this
  fetchData(article_id) {
    return actions.getArticleChangeRequest(article_id);
  }

  handleSave = e => {
    const id = this.state.changeRequest[0]._id;
    const accept = e.target.name === 'accept';
    actions
      .saveArticleChangeRequest({ id, accept })
      .then(json => {
        console.log('redirect to change list?', json);
      })
      .catch(() => {
        this.setState({
          message: {
            show: true,
            error: true,
            text: 'Save Unsuccessful',
          },
        });
      });
  };

  handleClose = () => {
    this.setState(() => ({
      message: { ...this.state.message, show: false },
    }));
  };

  render() {
    const { message, horizontal, vertical, mobile } = this.state;
    const { changeRequest } = this.state;
    // if we have an id but no topics we're not ready
    if (changeRequest.length === 0) {
      return <div>Loading ...</div>;
    }
    return (
      <Wrapper mobile={mobile}>
        <Heading1>Content Moderation</Heading1>
        <div>
          <div>Title: {changeRequest[0].article.title}</div>
          <div>Topic: {changeRequest[0].article.topic.name}</div>
          <div>SubTopic: {changeRequest[0].article.sub_topic.name}</div>
        </div>
        <div>
          {changeRequest[0].diff.map((part, ndx) => {
            const style = {
              // eslint-disable-next-line no-nested-ternary
              color: part.added ? 'green' : part.removed ? 'red' : 'grey',
            };
            return (
              // eslint-disable-next-line react/no-array-index-key
              <span key={ndx} style={style}>
                {part.value}
              </span>
            );
          })}
        </div>
        <div>
          <PrimaryButton name="accept" onClick={this.handleSave}>
            Accept
          </PrimaryButton>
          <PrimaryButton name="reject" onClick={this.handleSave}>
            Reject
          </PrimaryButton>
        </div>
        <MessageBar
          anchor={{ vertical, horizontal }}
          message={message}
          handleClose={this.handleClose}
        />
      </Wrapper>
    );
  }
}
