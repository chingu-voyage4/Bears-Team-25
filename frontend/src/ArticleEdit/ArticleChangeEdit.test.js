import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import ArticleChangeEdit from './ArticleChangeEdit';

const request = {
  _id: '5b133cf0ebd521b1d4a5144a',
  article: {
    id: '5ae27568463f344afc3ac08e',
    title: 'test article',
    topic: { id: 'topic_id', name: 'test topic' },
    sub_topic: { id: 'sub_topic_id', name: 'test sub topic' },
    content: 'TBD',
  },
  id: '5ae27568463f344afc3ac08e',
  title: 'test article',
  topic: { id: 'topic_id', name: 'test topic' },
  sub_topic: { id: 'sub_topic_id', name: 'test sub topic' },
  diff: [
    {
      count: 3,
      value: 'TBD',
    },
    {
      count: 5,
      added: true,
      value: ' test',
    },
  ],
  status: 'pending',
  requester: { id: '5ae34a775cfd8b560525c60c', username: 'test user' },
};

const props = {
  history: {
    push: jest.fn(),
  },
  location: {
    pathname: '/requests',
  },
};

describe('Article change request list', () => {
  it('should match snapshot', () => {
    jest
      .spyOn(ArticleChangeEdit.prototype, 'fetchData')
      .mockImplementation(() => new Promise(resolve => resolve(request)));

    const comp = shallow(
      <ArticleChangeEdit {...props} id="5b133cf0ebd521b1d4a5144a" />
    );
    return Promise.resolve().then(() => {
      comp.update();
      expect(toJSON(comp)).toMatchSnapshot();
    });
  });
});
