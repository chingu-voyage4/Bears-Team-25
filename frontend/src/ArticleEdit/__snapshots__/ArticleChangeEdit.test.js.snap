// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article change request list should match snapshot 1`] = `
<styled.div>
  <Heading1>
    Content Moderation
  </Heading1>
  <WithStyles(List)>
    <WithStyles(ListItem)>
      <Label>
        <WithStyles(Typography)
          variant="title"
        >
          Title:
        </WithStyles(Typography)>
      </Label>
      <Label>
        <WithStyles(Typography)
          variant="subheading"
        >
          test article
        </WithStyles(Typography)>
      </Label>
    </WithStyles(ListItem)>
    <WithStyles(ListItem)>
      <Label>
        <WithStyles(Typography)
          variant="title"
        >
          Topic:
        </WithStyles(Typography)>
      </Label>
      <Label>
        <WithStyles(Typography)
          variant="subheading"
        >
          test topic
        </WithStyles(Typography)>
      </Label>
    </WithStyles(ListItem)>
    <WithStyles(ListItem)>
      <Label>
        <WithStyles(Typography)
          variant="title"
        >
          Sub Topic:
        </WithStyles(Typography)>
      </Label>
      <Label>
        <WithStyles(Typography)
          variant="subheading"
        >
          test sub topic
        </WithStyles(Typography)>
      </Label>
    </WithStyles(ListItem)>
  </WithStyles(List)>
  <Styled(WithStyles(Paper))>
    <WithStyles(Paper)>
      <span
        key="0"
        style={
          Object {
            "color": "grey",
          }
        }
      >
        TBD
      </span>
      <span
        key="1"
        style={
          Object {
            "color": "green",
          }
        }
      >
         test
      </span>
    </WithStyles(Paper)>
  </Styled(WithStyles(Paper))>
  <styled.div>
    <PrimaryButton
      name="accept"
      onClick={[Function]}
    >
      Accept
    </PrimaryButton>
    <PrimaryButton
      name="reject"
      onClick={[Function]}
    >
      Reject
    </PrimaryButton>
  </styled.div>
  <MessageBar
    anchor={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    handleClose={[Function]}
    message={
      Object {
        "error": false,
        "show": false,
        "text": "",
      }
    }
  />
</styled.div>
`;
